src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -385,7 +407,62 @@ Status GcsActorManager::RegisterActor(
     return Status::OK();
   }
 
-  auto actor = std::make_shared<GcsActor>(request);
+  // Mark the callback as pending and invoke it after the actor has been successfully
+  // created.
+  actor_to_register_callbacks_[actor_id].emplace_back(std::move(callback));
+
+  auto actor = std::make_shared<GcsActor>(request.task_spec());
+  RAY_CHECK(registered_actors_.emplace(actor->GetActorID(), actor).second);
+
+  const auto &owner_address = actor->GetOwnerAddress();
+  auto node_id = ClientID::FromBinary(owner_address.raylet_id());
+  auto worker_id = WorkerID::FromBinary(owner_address.worker_id());
+  RAY_CHECK(unresolved_actors_[node_id][worker_id].emplace(actor->GetActorID()).second);
+
+  return gcs_table_storage_->ActorTable().Put(
+      actor->GetActorID(), *actor->GetMutableActorTableData(),
+      [this, actor](const Status &status) {
+        RAY_CHECK_OK(status);
+        // Invoke all callbacks for all registration requests of this actor (duplicated
+        // requests are included) and remove all of them from
+        // actor_to_register_callbacks_.
+        auto iter = actor_to_register_callbacks_.find(actor->GetActorID());
+        RAY_CHECK(iter != actor_to_register_callbacks_.end() && !iter->second.empty());
+        for (auto &callback : iter->second) {
+          callback(actor);
+        }
+        actor_to_register_callbacks_.erase(iter);
+      });
+}
+
+Status GcsActorManager::ReportActorDependenciesResolved(
+    const ray::rpc::ReportActorDependenciesResolvedRequest &request,
+    ReportActorDependenciesResolvedCallback callback) {
+  RAY_CHECK(callback);
+  const auto &actor_creation_task_spec = request.task_spec().actor_creation_task_spec();
+  auto actor_id = ActorID::FromBinary(actor_creation_task_spec.actor_id());
+
+  auto iter = registered_actors_.find(actor_id);
+  if (iter != registered_actors_.end() &&
+      iter->second->GetState() == rpc::ActorTableData::ALIVE) {
+    // When the network fails, Driver/Worker is not sure whether GcsServer has received
+    // the request of actor creation task, so Driver/Worker will try again and again until
+    // receiving the reply from GcsServer. If the actor has been created successfully then
+    // just reply to the caller.
+    callback(iter->second);
+    return Status::OK();
+  }
+
+  auto pending_register_iter = actor_to_register_callbacks_.find(actor_id);
+  if (pending_register_iter != actor_to_register_callbacks_.end()) {
+    // It is a duplicate message, just mark the callback as pending and invoke it after
+    // the actor has been successfully created.
+    pending_register_iter->second.emplace_back(std::move(callback));
```

Comment:
This works only when CompletePendingTask and PendingTaskFailed are idempotent I believe? @stephanie-wang Are they?